[WIP] Return exceptions in feeds

logmanoriginal opened the pull request:
This PR is work in progress, do not merge!

Changes exception handling in order to return regular feed items for exceptions happening during bridge execution. That way feed readers will display errors as if the bridge was updated regularly.

References #814, #828

Ping @ORelio 

### TODO
 - [x] Generate unique IDs to prevent feed readers from hiding re-occurring errors
 - [x] Include unique ID in the title to prevent feed readers from "updating" existing feeds based on names
 - [x] Add caching for errors (24 h)

logmanoriginal committed:
[Exceptions] Don't return header for bridge exceptions

logmanoriginal committed:
[Exceptions] Add link to list in exception message

This is an alternative when the button is not rendered
for some reason.

logmanoriginal committed:
[index] Don't return bridge exception for formats

logmanoriginal committed:
[index] Return feed item for bridge exceptions

logmanoriginal commented:
Some impressions

`?action=display&bridge=ABCTabs&format=Atom`
![image](https://user-images.githubusercontent.com/5776685/45836919-58b13380-bd0e-11e8-811c-6d663d5d3319.png)

`?action=display&bridge=ABCTabs&format=Html`
![image](https://user-images.githubusercontent.com/5776685/45836962-767e9880-bd0e-11e8-8cd4-7001ecec9e14.png)

Example of a format error
![image](https://user-images.githubusercontent.com/5776685/45837090-ca897d00-bd0e-11e8-9c41-05537350cb69.png)

mitsukarenai reviewed (approved):
seems good to me 👍

ORelio commented:
The result seems good :+1:

Some remarks about URL generation:

````php
$item['uri'] = http_build_query($params);
````
Wouldn't this cause the problem described [here](https://github.com/RSS-Bridge/rss-bridge/issues/814#issuecomment-421154422)? i.e. same URL whenever an error occur so only the first error will appear in feed reader for each bridge URL, once fixed, subsequent errors won't appear in the reader.

````php
$format->setLastModified($ctime);
````
Does this value control the cache duration? If so, maybe an error could generate a random ID by appending a random string to URL, then set `$ctime` to 24 hours?

logmanoriginal commented:
>Wouldn't this cause the problem described here?

Yes, absolutely. I haven't implemented unique IDs yet.

>Does this value control the cache duration?

No, the cache duration is managed inside the bridge (caching is currently missing for errors). This value indicates the modification date.

The returned value is provided to the format

https://github.com/RSS-Bridge/rss-bridge/blob/eb30038d6b267818cf951d91c265e205126bd98b/lib/FormatAbstract.php#L31-L33

and later used for the header

https://github.com/RSS-Bridge/rss-bridge/blob/eb30038d6b267818cf951d91c265e205126bd98b/lib/FormatAbstract.php#L39-L46

The variable name is based on the function used to receive the modified date from the bridge

https://github.com/RSS-Bridge/rss-bridge/blob/eb30038d6b267818cf951d91c265e205126bd98b/lib/BridgeAbstract.php#L296-L298

The function name is quite misleading and not to confuse with [`getCacheTimeout`](https://github.com/RSS-Bridge/rss-bridge/blob/eb30038d6b267818cf951d91c265e205126bd98b/lib/BridgeAbstract.php#L292-L294), which is what you were thinking about. I think I'll actually change the function name to prevent confusion in the future.

logmanoriginal committed:
[BridgeAbstract] Rename 'getCacheTime' to 'getModifiedTime'

logmanoriginal committed:
[BridgeAbstract] Move caching to index.php to separate concerns

index.php needs more control over caching behavior in order to cache
exceptions. This cannot be done in a bridge, as the bridge might be
broken, thus preventing caching from working.

This also (and more importantly) separates concerns. The bridge should
not even care if caching is involved or not. Its purpose is to collect
and provide data.

Response times should be faster, as more complex bridge functions like
'setDatas' (evaluates all input parameters to predict the current
context) and 'collectData' (collects data from sites) can be skipped
entirely.

Notice: In its current form, index.php takes care of caching. This
could, however, be moved into a separate class (i.e. CacheAbstract)
in order to make implementation details cache specific.

logmanoriginal committed:
[index] Add '_error_time' parameter to $item['uri']

This ensures that error messages are recognized by feed readers as
new errors after 24 hours. During that time the same item is returned
no matter how often the cache is cleared.

References https://github.com/RSS-Bridge/rss-bridge/issues/814#issuecomment-420876162

logmanoriginal committed:
[index] Include '_error_time' in the title for errors

This prevents feed readers from "updating" feeds based on the title

logmanoriginal commented:
More commits added. This should be pretty stable by now, but needs some further testing.

Notice: `BridgeAbstract` no longer handles caching, this is now done in `index.php` for reasons mentioned in the commit message. Also makes it much more readable in my opinion. A `CacheAbstract` could be necessary to reduce the amount of code in `index.php` though.

logmanoriginal committed:
[index] Handle "HTTP_IF_MODIFIED_SINCE" client requests

Implementation is based on `BridgeAbstract::dieIfNotModified()`,
introduced in 422c125d8e25b9ac5209580b1976fb5d75dab8f8 and
simplified based on https://stackoverflow.com/a/10847262

Basically, before returning cached data we check if the client send
the "HTTP_IF_MODIFIED_SINCE" header. If the modification time is
more recent or equal to the cache time, we reply with "HTTP/1.1 304
 Not Modified" (same as before). Otherwise send the cached data.

logmanoriginal committed:
[index] Don't encode exception message with `htmlspecialchars`

logmanoriginal committed:
[Exceptions] Include error message in exception

logmanoriginal committed:
[index] Show different error message for error code 0

logmanoriginal commented:
Rebased on master. Seems to be working well on my end, so it could be merged. Any objections?
